lib/slog/slog.go

```diff
@@ -0,0 +1,125 @@
+	Error      error          `json:"error,omitempty"`      // Error is an optional error
+	Details    any            `json:"details,omitempty"`    // Details are optional details
+	StackTrace string         `json:"stacktrace,omitempty"` // StackTrace is optional stack trace
+	ClientID   *core.ClientID `json:"clientid,omitempty"`   // ClientID is optional id of client, if known.
+	Upstream   *core.Upstream `json:"upstream,omitempty"`   // Upstream is optional upstream, if known.
+}
+
+// Logger is an abstract log interface for the server.
+//
+// Multiple goroutines may invoke methods on a Logger simultaneously.
+type Logger interface {
+	Info(record *LogRecord)
+	Warn(record *LogRecord)
+	Error(record *LogRecord)
+}
+
+// TODO make the log output less awful to read by humans and machines.
+type stdlibLogShim struct{}
+
+type errorPayload struct {
+	Type  string `json:"type,omitempty"`  // Type is the error type
+	Error string `json:"error,omitempty"` // Error is the error message
+}
+
+func asErrorPayload(err error) *errorPayload {
+	if err == nil {
+		return nil
+	}
+	return &errorPayload{
+		Type:  fmt.Sprintf("%T", err),
+		Error: err.Error(),
+	}
+}
+
+type recordPayload struct {
+	Msg        string         `json:"msg,omitempty"`        // Msg is an optional log message
+	Error      *errorPayload  `json:"error,omitempty"`      // Error is an optional error
+	Details    any            `json:"details,omitempty"`    // Details are optional details
+	StackTrace string         `json:"stacktrace,omitempty"` // StackTrace is optional stack trace
+	ClientID   *core.ClientID `json:"clientid,omitempty"`   // ClientID is optional id of client, if known.
+	Upstream   *core.Upstream `json:"upstream,omitempty"`   // Upstream is optional upstream, if known.
+	Level      string         `json:"level,omitempty"`
+}
+
+func logRecordAsSemiJSON(level string, record *LogRecord) {
+	var payload recordPayload
+	payload.Level = level
+	if record != nil {
+		payload.Msg = record.Msg
+		payload.Error = asErrorPayload(record.Error)
+		payload.Details = record.Details
+		payload.StackTrace = record.StackTrace
+		payload.ClientID = record.ClientID
+		payload.Upstream = record.Upstream
+	}
+
+	data, _ := json.Marshal(&payload)
+
+	// TODO put the timestamps in the JSON as well.
+	log.Println(string(data))
+}
+
+func (s *stdlibLogShim) Info(record *LogRecord) {
+	logRecordAsSemiJSON(InfoLevel, record)
+}
+
+func (s *stdlibLogShim) Warn(record *LogRecord) {
+	logRecordAsSemiJSON(WarnLevel, record)
+}
+
+func (s *stdlibLogShim) Error(record *LogRecord) {
+	logRecordAsSemiJSON(ErrorLevel, record)
+}
+
+// GetDefaultLogger returns the default Logger.
+func GetDefaultLogger() Logger {
+	return &stdlibLogShim{}
+}
+
+// RecordingLogger captures all logged events in memory.
+// It is designed for use as a test fixture.
+type RecordingLogger struct {
+	Events []Event
+}
+
+type Event struct {
+	Level string
+	*LogRecord
+}
+
+func (l *RecordingLogger) Info(record *LogRecord) {
```

Comment:
This is a somewhat heavy-handed interface compared to most log APIs. Not a huge deal, but you'd have to rewrite pretty much every call to move to something else.

Reply:
I agree.   This API is not a good idea, and to fix it you'd need to rewrite each call.

It felt like a good idea for about 5 minutes "i can write a simple structured logging interface, then i can easily test stuff about logging without needing to parse through strings", but the API design was about 1 minute of not very clear thought.

Reading the calls, it is obviously not a winning idea:

`logger.Info(&slog.LogRecord{Msg: fmt.Sprintf(` vs `log.Printf(`